internal/monitors/elasticsearch/elasticsearch.go

```diff
@@ -0,0 +1,414 @@
+
+	if err != nil {
+		logger.WithError(err).Errorf("Failed to determine whether current node is the master")
+		return err
+	}
+
+	clusterName := nodeInfoOutput.ClusterName
+	sinfo.defaultDimensions, err = prepareDefaultDimensions(configuredClusterName, clusterName)
+
+	if err != nil {
+		logger.WithError(err).Errorf("Failed to prepare plugin_instance dimension")
+		return err
+	}
+
+	sinfo.nodeMetricDimensions = utils.MergeStringMaps(sinfo.defaultDimensions, nodeDimensions)
+
+	return nil
+}
+
+// Returns all fields of a shared info object
+func (sinfo *sharedInfo) getAllSharedInfo() (map[string]string, map[string]string, bool) {
+	sinfo.lock.RLock()
+	defer sinfo.lock.RUnlock()
+
+	return sinfo.defaultDimensions, sinfo.nodeMetricDimensions, sinfo.nodeIsCurrentMaster
+}
+
+// Configure monitor
+func (m *Monitor) Configure(conf *Config) error {
+	esClient := client.NewESClient(conf.Host, conf.Port, conf.UseHTTPS, conf.Username, conf.Password)
+	m.ctx, m.cancel = context.WithCancel(context.Background())
+
+	// To handle metadata about metrics that is shared
+	shared := sharedInfo{
+		lock:          &sync.RWMutex{},
+		isInitialized: false,
+	}
+
+	utils.RunOnInterval(m.ctx, func() {
+		// Fetch metadata from Elasticsearch nodes
+		err := shared.fetchNodeAndClusterMetadata(esClient, conf.Cluster)
+
+		// For any reason the monitor is not able to fetch cluster and node information upfront we capture that
+		// to ensure that the stats are not sent in with faulty dimensions. The monitor will try to fetch this
+		// information again every MetadataRefreshInterval seconds
+		if err != nil {
+			logger.WithError(err).Errorf(fmt.Sprintf("Failed to get Cluster and node metadata upfront. Will try again in %d s.", conf.MetadataRefreshIntervalSeconds))
+			return
+		}
+
+		// The first time fetchNodeAndClusterMetadata succeeded the monitor can schedule fetchers for all the other stats
+		if !shared.isInitialized {
+
+			// Collect Node level stats
+			utils.RunOnInterval(m.ctx, func() {
+				_, defaultNodeMetricDimensions, _ := shared.getAllSharedInfo()
+
+				m.fetchNodeStats(esClient, conf, defaultNodeMetricDimensions)
+			}, time.Duration(conf.IntervalSeconds)*time.Second)
+
+			// Collect cluster level stats
+			utils.RunOnInterval(m.ctx, func() {
+				defaultDimensions, _, nodeIsCurrentMaster := shared.getAllSharedInfo()
+
+				if !*conf.ClusterHealthStatsMasterOnly || nodeIsCurrentMaster {
+					m.fetchClusterStats(esClient, conf, defaultDimensions)
+				}
+			}, time.Duration(conf.IntervalSeconds)*time.Second)
+
+			// Collect Index stats
+			utils.RunOnInterval(m.ctx, func() {
+				defaultDimensions, _, nodeIsCurrentMaster := shared.getAllSharedInfo()
+
+				// if "IndexStatsMasterOnly" is true collect stats only from master,
+				// otherwise collect index stats from all nodes
+				if *conf.EnableIndexStats && (!*conf.IndexStatsMasterOnly || nodeIsCurrentMaster) {
```

Comment:
Don't even schedule this function if `*conf.EnableIndexStats` is `false`.  Config is immutable so you don't have to worry about it changing.

Reply:
Updated so that we only check this once in configure